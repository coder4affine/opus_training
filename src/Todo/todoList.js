import React from 'react';
import PropTypes from 'prop-types';

const todoList = ({
  todoItem,
  updatingTodo,
  updateText,
  changeTodoText,
  markDone,
  updateTodo,
  deleteTodo,
}) => {
  return (
    <div
      key={todoItem.id}
      style={{
        display: 'flex',
        padding: 20,
        width: '100%',
      }}
    >
      <input type="checkbox" checked={todoItem.isDone} onChange={() => markDone(todoItem)} />
      {updatingTodo && todoItem.id === updatingTodo.id ? (
        <input
          type="text"
          value={updatingTodo.text}
          onChange={updateText}
          onBlur={changeTodoText}
        />
      ) : (
        <span
          style={{
            textDecoration: todoItem.isDone ? 'line-through' : 'none',
            flex: 1,
          }}
        >
          {todoItem.text}
        </span>
      )}
      <button type="button" onClick={() => updateTodo(todoItem)}>
        Update
      </button>

      <button type="button" onClick={() => deleteTodo(todoItem)}>
        Delete
      </button>
    </div>
  );
};

todoList.propTypes = {
  todoItem: PropTypes.shape({
    id: PropTypes.number.isRequired,
    isDone: PropTypes.bool.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  updatingTodo: PropTypes.shape({
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
  }),
  updateText: PropTypes.func.isRequired,
  changeTodoText: PropTypes.func.isRequired,
  markDone: PropTypes.func.isRequired,
  updateTodo: PropTypes.func.isRequired,
  deleteTodo: PropTypes.func.isRequired,
};

todoList.defaultProps = {
  updatingTodo: null,
};

export default todoList;
