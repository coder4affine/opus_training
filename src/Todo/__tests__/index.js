import React from 'react';

import Index from '../index';
import TodoForm from '../todoForm';
import TodoStatus from '../todoStatus';
import todoList from '../todoList';

function setup() {
  const props = {};
  const Wrapper = shallow(<Index />);
  return {
    props,
    Wrapper,
  };
}

describe('snapshot', () => {
  it('snapshot of index', () => {
    const { Wrapper } = setup();
    expect(Wrapper).toMatchSnapshot();
  });
});

describe('Check index page view', () => {
  it('description', () => {
    const { Wrapper } = setup();
    const h1 = Wrapper.find('h1');
    expect(h1.text()).toEqual('Todo App');

    const errorMessage = 'oops! something went Wrong';
    Wrapper.setState({ error: new Error(errorMessage) });
    const p2 = Wrapper.find('p');
    expect(p2.text()).toEqual(errorMessage);
  });

  it('check p tag exist', () => {
    const { Wrapper } = setup();
    Wrapper.setState({ error: null });
    const p1 = Wrapper.find('p');
    expect(p1.exists()).toBe(false);
  });

  it('check todo form exist with props', () => {
    const { Wrapper } = setup();
    const form = Wrapper.find(TodoForm);
    expect(form.exists()).toBe(true);
    expect(form.props()).toEqual({
      addTodo: expect.any(Function),
      todoText: '',
      changeText: expect.any(Function),
    });
  });

  it('TodoStatus exist', () => {
    const { Wrapper } = setup();
    const statusComponent = Wrapper.find(TodoStatus);
    expect(statusComponent.exists()).toBe(true);
    expect(statusComponent.props()).toEqual({
      changeStatus: expect.any(Function),
      status: 'all',
    });
  });

  it('TodoStatus exist', () => {
    const { Wrapper } = setup();
    const statusComponent = Wrapper.find(TodoStatus);
    expect(statusComponent.exists()).toBe(true);
    expect(statusComponent.props()).toEqual({
      changeStatus: expect.any(Function),
      status: 'all',
    });
  });

  it('Todo List Exist', () => {
    const { Wrapper } = setup();
    const listComponent = Wrapper.find(todoList);
    expect(listComponent.exists()).toBe(false);

    const todoItem = {
      id: 1,
      text: 'get milk',
      isDone: false,
    };

    Wrapper.setState({
      todoList: [todoItem],
    });
    const listComponent1 = Wrapper.find(todoList);
    expect(listComponent1.exists()).toBe(true);
    expect(listComponent1.length).toBe(1);
    // expect(listComponent1.props()).toEqual({
    //   key: todoItem.id,
    //   todoItem,
    //   updatingTodo: null,
    //   updateText: Wrapper.instance.updateText(),
    //   changeTodoText: expect.any(Function),
    //   markDone: expect.any(Function),
    //   updateTodo: expect.any(Function),
    //   deleteTodo: expect.any(Function),
    // });
  });
});
