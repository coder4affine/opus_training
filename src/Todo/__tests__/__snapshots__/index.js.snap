// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot snapshot of index 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "margin": 0,
      "padding": 0,
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <h1>
      Todo App
    </h1>
    <p>
      fetch is not defined
    </p>
    <todoForm
      addTodo={[Function]}
      changeText={[Function]}
      todoText=""
    />
  </div>
  <todoStatus
    changeStatus={[Function]}
    status="all"
  />
</div>
`;
